### ISSUE TITLE
[New rule] Notify that foreach may throw InvalidCastException 

### ISSUE THREAD

**Issue by maxkoshevoi** [dotnet/roslyn#54712]
### Describe the problem you are trying to solve

As described [here](https://www.jetbrains.com/help/rider/PossibleInvalidCastExceptionInForeachLoop.html), [here](https://rules.sonarsource.com/csharp/RSPEC-3217), and [here](https://stackoverflow.com/a/949835/4312132) `foreach` can throw `InvalidCastException` at runtime because of hidden explicit cast that is does under the hood.

This is very dangerous behavior for statically typed language to have since programmers don't expect successfully compiled code to throw `InvalidCastException` when they didn't do any casting.
I came across this behavior when I changed type of a collection to more generic one and then got this error at runtime.

This behavior was needed until generics appeared in C# 2, but it's still here with us 15 years after it's introduction.

### Describe suggestions on how to achieve the rule

While solution could be as easy as "always use `var` for `foreach` items", I'm not a fan of this approach. `var` makes code less readable in this scenario, and forcing someone to use `var` is not in the spirit of this keyword anyway (it was made to be optional).

I'm proposing analyzer that would detect that instance of type `T` from `IEnumerable<T>` that is passed to `foreach` can be cast to type specified for `item` of the `foreach`.

Something like this, but using [IsAssignableFrom](https://docs.microsoft.com/en-us/dotnet/api/system.type.isassignablefrom?view=net-5.0):
```cs
List<A> aList = new() { new() };
Foreach<B, A>(aList);

List<B> bList = new() { new() };
Foreach<A, B>(bList);

void Foreach<TItem, TCollectionValue>(IEnumerable<TCollectionValue> collection) where TCollectionValue : TItem
{
    foreach (TItem item in collection)
    {
    }
}

class A { }
class B : A { }
```
![image](https://user-images.githubusercontent.com/6609929/100783634-b4ab9280-3416-11eb-92d9-364b307f2043.png)

### Additional context

Here is code to reproduce the issue:
```cs
List<A> aList = new() { new() };
foreach (B item in aList)
{
}

class A { }
class B : A { }
```
![image](https://user-images.githubusercontent.com/6609929/100781552-d9523b00-3413-11eb-9ac4-ea41b8f50679.png)

### Timeline

For anyone new here, this issue went quite a journey. Here's a timeline of the things that have happened:
- This issue got created (got closed based on discussion with runtime team №1, and reopened based on discussion with runtime team №2)
- PR that implements the analyzer: https://github.com/dotnet/roslyn-analyzers/pull/4868 (got closed based on discussion with runtime team №1)
- Discussion with runtime team https://github.com/dotnet/runtime/issues/48529 (they suggested to move it in StyleCop, and closed the issue)
- Moved the issue to StyleCop: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3307 (got rejected since it's not a style rule, since fix changes behavior)
- Reopened discussion with runtime folks (in a new issue): https://github.com/dotnet/runtime/issues/54629 (as a result this issue got reopened)

**Comment by RikkiGibson** [comment]
Thanks for your persistence in exploring this issue @maxkoshevoi. It looks like once you excluded collections with an `object Current`  and excluded things which do not implement `IEnumerable<T>` that we actually got 0 warnings when building the runtime. Is that correct?

**Comment by maxkoshevoi** [comment]
> Referencing the "issue timeline" to help those trying to get full context

Good point. Added `Timeline` to issue description

> it looks like once you excluded collections with an `object Current` and excluded things which do not implement `IEnumerable<T>` that we actually got 0 warnings when building the runtime

Looks correct:
- All matches: 933
- Without object collections: 65
- Without object and single type collections: 0

**Comment by srikrishnag** [comment]
Can we have this configurable rule with roslyn, with by default not enabled, and consumers can enabled it as per the needs ?

**Comment by CyrusNajmabadi** [comment]
@srikrishnag that's what is being discussed :)

**Comment by RikkiGibson** [comment]
Reiterating the suggestion that we would only give this warning if both of the above conditions are met:

1. The type of `Current` in the enumerator used in `foreach (var item in collection)` is *not* `object`.
2. `collection` implements `IEnumerable<T>`.

I would like to investigate more closely whether if we impose the above conditions before giving a warning, we achieve the following goals:

1. The overwhelming majority of the time the warning occurs in "real world" code, it means there is a real bug in the program. Think 99.9% of the time.
2. The warning *does* occur in enough of the situations where a problematic foreach variable downcast is found in user code, that it actually alleviates the need for authors/reviewers to manually check the type of the collection being enumerated against the foreach variable type.

If we are able to determine that both of the above goals are met when introducing the conditions, I would advocate for shipping the analyzer and making it a warning by default. FWIW, this analyzer would have prevented at least [one](https://github.com/dotnet/roslyn/pull/56233/files#diff-fd41468c64dd070639cc4e760f991439fbfe752874a14791d981ca127945d7fa) bug in Roslyn.

Maybe we could try this by packaging up @maxkoshevoi's analyzer implementation, then adding the reference to various existing projects that already build with analyzers. Clearly we've already done it with runtime, perhaps running on a couple more repos like roslyn or aspnetcore would help build evidence here.

**Comment by CyrusNajmabadi** [comment]
FYI, i'm working on a fixer here as well.

**Comment by RikkiGibson** [comment]
Should this be closed now that the analyzer/fixer is merged into 17.3?
